keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,540 @@
++ Ensure bucket consuming pods wait until the target buckets have been created and are accessible. Thus there is no specific order required for when a bucket claim is created vs. when the app pod is run.
++ Present similar user and admin experiences for both _greenfield_ (new) and _brownfield_ (existing) bucket provisioning.
+
+### Non-Goals
+
++ Update the native Kubernetes PVC-PV API to support object buckets.
++ Define a native _data-plane_ object store API.
++ Handle the small percentage of apps that will not be portable due to use of non-compatible object-store features.
+
+## Proposal
+
+Create a bucket provisioning library, similar to the Kubernetes [sig-storage-lib-external-provisioner](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/controller.go) library, which handles all of the controller and Kubernetes resources related to provisioning.
+This library will implement a simple and familiar API via Custom Resources, and define a _contract_ between app developers and provisioners regarding the Kubernetes resources related to bucket provisioning.
+
+The actual creation of physical buckets and the generation of appropriate credentials belong to each object store provisioner, while the library handles watches on bucket claims, reconciles desired state, and creates and deletes k8s resources.
+
+Since certain object store-specific artifacts, e.g. user name, are needed in order to clean up after a bucket is deleted or revoked, this proposal must allow for some provisioner values to be stored across API calls.
+
+## Design Details
+
+### Custom Resources
+
+(see the [full specs](#custom-resource-details))
+
+The bucket provisioning library utilizes two Custom Resources to abstract an object store bucket and a claim/request for such a bucket.
++ a namespaced _**ObjectBucketClaim**_ (OBC), similar to a PVC, is a formal request for a new or an existing bucket.
++ an _**ObjectBucket**_ (OB), similar to a PV, is the k8s representation of the provisioned bucket and may contain object store-specific data needed to allow the bucket to be de-provisioned. The OB resides in the same namespace as the object store, and is typically not seen by bucket consumers.
+**Note:** namespaced OBs is a change from the original design and was suggested at the 2019 NA Kubecon storage face-to-face with the premise being that if there is no _technical_ reason for a cluster scoped resource then it should be namespaced.
+
+As is true for PVCs-PVs, there is a 1:1 relationship between an OBC and an OB, and as will be seen below, there is a [_binding_](#binding) between an OBC and OB. 
+
+OBCs reference a storage class. The storage class references the external provisioner, defines a reclaim policy, and specifies object-store specific parameters, such as region, owner secret, bucket lifecycle policies, etc.
```

Comment:
Per previous comment, I think it would be better to have an ObjectBucketClass to handle these issues rather than overloading StorageClass with functionality not related to volumes.